playbooks/fundamentals/template_development_guide.adoc

```diff
@@ -0,0 +1,391 @@
+  displayName: Application Name
+  name: APPLICATION_NAME
+  required: true
+  value: openjdk-app
+...
+```
+
+=== Methods for Writing or Generating Templates
+
+==== Start from an existing template
+
+Exporting and modifying an existing template is many times the fastest path to success. Simply peruse through the set of templates provided out of the box by OpenShift, find the one closest to what you need, and export it.
+
+```
+$ oc get templates -n openshift
+...
+s2i-spring-boot-camel-config                    Spring Boot and Camel using ConfigMaps and Secrets. This quickstart demonstra...   13 (2 blank)      3
+...
+
+$ oc export template/s2i-spring-boot-camel-config -n openshift > my-new-spring-template.yml
+```
+
+Once exported the first thing to do is make sure to rename it. Just make sure and be thorough, a templates name is generally used multiple times in the template.
+
+```
+$ grep 's2i-spring-boot-camel-config\|my-new-spring-template' ./my-new-spring-template.yml
+  template: s2i-spring-boot-camel-config
+  name: s2i-spring-boot-camel-config
+  value: s2i-spring-boot-camel-config
+
+$ sed -i 's/s2i-spring-boot-camel-config/my-new-spring-template/g' ./my-new-spring-template.yml
+
+$ grep 's2i-spring-boot-camel-config\|my-new-spring-template' ./my-new-spring-template.yml
+  template: my-new-spring-template
+  name: my-new-spring-template
+  value: my-new-spring-template
+```
+
+From here, you're free to modify whatever needs modifying to meet your needs. When modifying an existing template, be aware that there is a lot of metadata in the form of labels and annotations that may or may not be relevant to your new template. The good news is that, if you are writing a template for automation purposes, and not for use in the Web Console, much of that stuff can be cleaned out, as it is mostly used to populate parts of the UI and little else. Just keep in mind that you may want to spend the time updating those values if you plan to create new Web Console quickstarts.
+
+==== Build from Scratch
+
+A more barebones approach is to simply write the template from scratch. This is especially nice when you need a very minimal template, and you want to keep it clean of any leftover metadata from the original template. Just start with this skeleton and you'll be good to go.
+
+```
+apiVersion: v1
+kind: Template
+labels:
+message: <Creation message>
+metadata:
+  name: <template name>
+objects:
+parameters:
+```
+
+==== Export existing objects as a Template
+
+Maybe the most powerful mode of creating a new template is to use `oc export` to generate one from a set of already created objects. This allows you to first build and wire up and application manually using the client tools and/or the Web Console, and then capture your work in the form of a repeatable template.
+
+=== Best Practices for Template Writing
+
+The following is a list of suggested best practices for template writing.
+
+- **Include a `template` label in all objects.**
```

Comment:
For labels to all objects, `oc apply` has a `-l` option to add a label to all objects created.  Web Console template wizard also has similar functionality. Makes sense if you want to advocate hard-coding in the template objects; it just leaves room for human error.